sdk/iot/hub/src/az_iot_hub_client_c2d.c

```diff
@@ -12,28 +12,38 @@
 
 static const az_span c2d_topic_prefix = AZ_SPAN_LITERAL_FROM_STR("devices/");
 static const az_span c2d_topic_suffix = AZ_SPAN_LITERAL_FROM_STR("/messages/devicebound/");
+static const uint8_t null_terminator = '\0';
 static const uint8_t hash_tag = '#';
 
 AZ_NODISCARD az_result az_iot_hub_client_c2d_subscribe_topic_filter_get(
     az_iot_hub_client const* client,
-    az_span mqtt_topic_filter,
-    az_span* out_mqtt_topic_filter)
+    char* mqtt_topic_filter,
+    size_t mqtt_topic_filter_size,
+    size_t* out_mqtt_topic_filter_length)
 {
   AZ_PRECONDITION_NOT_NULL(client);
-  AZ_PRECONDITION_VALID_SPAN(mqtt_topic_filter, 0, false);
-  AZ_PRECONDITION_NOT_NULL(out_mqtt_topic_filter);
+  AZ_PRECONDITION_NOT_NULL(mqtt_topic_filter);
+  AZ_PRECONDITION(mqtt_topic_filter_size > 0);
 
+  az_span mqtt_topic_filter_span
+      = az_span_init((uint8_t*)mqtt_topic_filter, (int32_t)mqtt_topic_filter_size);
   int32_t required_length = az_span_size(c2d_topic_prefix)
-      + az_span_size(client->_internal.device_id) + az_span_size(c2d_topic_suffix) + 1;
+      + az_span_size(client->_internal.device_id) + az_span_size(c2d_topic_suffix)
+      + (int32_t)sizeof(hash_tag);
```

Comment:
CALLOUT: Any individual `uint8_t` will have a call to `sizeof()` in the required size so that we can easily see which characters are accounted for and which are not. If we have a situation where 3+ characters are copied to the output and we just put a number here, it can be easy to lose track.